File: android/app/src/main/kotlin/com/torvm/android/dns/DnsRelay.kt
package com.torvm.android.dns

import com.torvm.android.packet.IPv4Header
import com.torvm.android.packet.PacketBuilder
import com.torvm.android.packet.UdpHeader
import com.torvm.android.tunnel.TunWriter
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress

/**
 * Relays DNS queries to an external DNS resolver (typically Tor's DNSPort)
 * and injects the responses back into the TUN device.
 *
 * Each query is forwarded as a plain UDP datagram. The [protector] lambda
 * must be invoked on the socket so that the VpnService marks it as
 * "protected" -- preventing the DNS traffic itself from being routed back
 * through the TUN and creating an infinite loop.
 */
class DnsRelay(
    private val dnsHost: String,
    private val dnsPort: Int,
    private val tunWriter: TunWriter,
    private val protector: ((DatagramSocket) -> Boolean)?
) {
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    companion object {
        /** Maximum size of a DNS UDP response. */
        private const val DNS_BUFFER_SIZE = 4096

        /** Socket timeout for upstream DNS queries. */
        private const val DNS_TIMEOUT_MS = 5000

        private const val TAG = "DnsRelay"
    }

    /**
     * Start the relay.
     *
     * The relay is stateless -- this method is a no-op provided for symmetry
     * with [stop].
     */
    fun start() {
        // No-op; DNS relay handles each packet independently
    }

    /**
     * Cancel all in-flight DNS queries.
     */
    fun stop() {
        scope.cancel()
    }

    /**
     * Handle an intercepted DNS query packet.
     *
     * The query [payload] (raw DNS wire format) is forwarded to the
     * configured resolver. On success the DNS response is wrapped in a
     * UDP/IP packet with source and destination swapped and written back
     * to the TUN device.
     */
    suspend fun handlePacket(
        ipHeader: IPv4Header,
        udpHeader: UdpHeader,
        payload: ByteArray
    ) {
        scope.launch {
            try {
                val dnsResponse = forwardDnsQuery(payload)

                val responsePacket = PacketBuilder.buildUdpPacket(
                    srcAddr = ipHeader.destAddressBytes,
                    dstAddr = ipHeader.sourceAddressBytes,
                    srcPort = udpHeader.destPort,
                    dstPort = udpHeader.sourcePort,
                    payload = dnsResponse
                )

                tunWriter.write(responsePacket)
            } catch (_: Exception) {
                // Query failed (timeout, network error, etc.) -- drop silently.
                // The application will retry or fall back to TCP DNS.
            }
        }
    }

    /**
     * Forward a raw DNS query to the upstream resolver and return the
     * response bytes.
     */
    private fun forwardDnsQuery(query: ByteArray): ByteArray {
        val socket = DatagramSocket()
        try {
            protector?.invoke(socket)
            socket.soTimeout = DNS_TIMEOUT_MS

            val address = InetAddress.getByName(dnsHost)
            socket.send(DatagramPacket(query, query.size, address, dnsPort))

            val buffer = ByteArray(DNS_BUFFER_SIZE)
            val response = DatagramPacket(buffer, buffer.size)
            socket.receive(response)

            return buffer.copyOf(response.length)
        } finally {
            socket.close()
        }
    }
}

File: ios/TorVMTunnel/DNS/DnsRelay.swift
import Foundation
import Network
import NetworkExtension
import os

/// Relays DNS queries to an external DNS resolver (typically Tor's DNSPort)
/// and injects the responses back into the TUN device.
///
/// Each query is forwarded as a plain UDP datagram via NWConnection.
/// On iOS, routing loop prevention is handled by NEIPv4Settings.excludedRoutes
/// rather than Android's socket protection pattern.
final class DnsRelay {

    private let dnsHost: String
    private let dnsPort: UInt16
    private let packetFlow: NEPacketTunnelFlow

    private static let dnsBufferSize = 4096
    private static let dnsTimeoutSeconds: TimeInterval = 5
    private static let maxConcurrentQueries = 32

    private let activeLock = os.OSAllocatedUnfairLock(initialState: Int(0))
    private let queue = DispatchQueue(label: "com.torvm.dns", attributes: .concurrent)
    private var isStopped = false

    init(dnsHost: String, dnsPort: UInt16, packetFlow: NEPacketTunnelFlow) {
        self.dnsHost = dnsHost
        self.dnsPort = dnsPort
        self.packetFlow = packetFlow
    }

    func start() {
        isStopped = false
    }

    func stop() {
        isStopped = true
    }

    /// Handle an intercepted DNS query packet.
    func handlePacket(ipHeader: IPv4Header, udpHeader: UdpHeader, payload: Data) {
        guard !isStopped else { return }

        let allowed = activeLock.withLock { (count: inout Int) -> Bool in
            guard count < Self.maxConcurrentQueries else { return false }
            count += 1
            return true
        }
        guard allowed else { return }

        queue.async { [weak self] in
            defer {
                self?.activeLock.withLock { (count: inout Int) in
                    count -= 1
                }
            }
            guard let self = self, !self.isStopped else { return }

            do {
                let response = try self.forwardQuerySync(payload)

                let responsePacket = PacketBuilder.buildUdpPacket(
                    srcAddr: ipHeader.destAddress,
                    dstAddr: ipHeader.sourceAddress,
                    srcPort: udpHeader.destPort,
                    dstPort: udpHeader.sourcePort,
                    payload: response
                )

                self.packetFlow.writePackets(
                    [responsePacket],
                    withProtocols: [NSNumber(value: AF_INET)]
                )
            } catch {
                // Query failed — drop silently.
            }
        }
    }

    /// Forward a raw DNS query to the upstream resolver using NWConnection (UDP).
    /// TODO: Replace synchronous semaphore-based NWConnection pattern with async/await
    /// using withCheckedThrowingContinuation for better thread utilization.
    private func forwardQuerySync(_ query: Data) throws -> Data {
        guard query.count >= 2 else {
            throw DnsRelayError.queryTooShort
        }

        let sem = DispatchSemaphore(value: 0)
        var result: Result<Data, Error>?

        let endpoint = NWEndpoint.hostPort(
            host: NWEndpoint.Host(dnsHost),
            port: NWEndpoint.Port(rawValue: dnsPort)!
        )
        let params = NWParameters.udp
        let connection = NWConnection(to: endpoint, using: params)

        connection.stateUpdateHandler = { state in
            switch state {
            case .ready:
                // Send query
                connection.send(content: query, completion: .contentProcessed { error in
                    if let error = error {
                        result = .failure(error)
                        sem.signal()
                        return
                    }
                    // Receive response
                    connection.receiveMessage { data, _, _, recvError in
                        defer { sem.signal() }
                        if let error = recvError {
                            result = .failure(error)
                            return
                        }
                        guard let data = data, data.count >= 2 else {
                            result = .failure(DnsRelayError.emptyResponse)
                            return
                        }
                        // Validate transaction ID
                        if data[0] != query[0] || data[1] != query[1] {
                            result = .failure(DnsRelayError.transactionIdMismatch)
                            return
                        }
                        result = .success(data)
                    }
                })
            case .failed(let error):
                result = .failure(error)
                sem.signal()
            case .cancelled:
                result = .failure(DnsRelayError.cancelled)
                sem.signal()
            default:
                break
            }
        }

        connection.start(queue: queue)

        let timeout = DispatchTime.now() + Self.dnsTimeoutSeconds
        if sem.wait(timeout: timeout) == .timedOut {
            connection.cancel()
            throw DnsRelayError.timeout
        }

        connection.cancel()

        switch result {
        case .success(let data):
            return data
        case .failure(let error):
            throw error
        case .none:
            throw DnsRelayError.emptyResponse
        }
    }

    enum DnsRelayError: LocalizedError {
        case queryTooShort
        case emptyResponse
        case transactionIdMismatch
        case timeout
        case cancelled

        var errorDescription: String? {
            switch self {
            case .queryTooShort: return "DNS query too short"
            case .emptyResponse: return "Empty DNS response"
            case .transactionIdMismatch: return "DNS transaction ID mismatch"
            case .timeout: return "DNS query timed out"
            case .cancelled: return "DNS query cancelled"
            }
        }
    }
}
